File: tests/python/dynamic_programming/longest_increasing_subsequence/test_longest_increasing_subsequence.py
from problems.dynamic_programming.longest_increasing_subsequence.python.longest_increasing_subsequence import \
    longestIncreasingSubsequence
from tests.python.utility import load_test_cases


def test_longest_increasing_subsequence() -> None:

    test_cases = load_test_cases(
        "dynamic_programming/longest_increasing_subsequence.json"
    )

    for idx, test in enumerate(test_cases):
        array: list[int] = test["array"]
        expected: list[int] = test["expected"]

        result: list[int] = longestIncreasingSubsequence(array)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/dynamic_programming/test_longest_increasing_subsequence.cpp
#include "../../../problems/dynamic_programming/longest_increasing_subsequence/cpp/longest_increasing_subsequence.h"
#include "../jsontestbase.h"

#include <vector>

class LongestIncreasingSubsequenceTest : public JsonTestBase {
public:
  LongestIncreasingSubsequenceTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/dynamic_programming/longest_increasing_subsequence.json";
  }
};

TEST_F(LongestIncreasingSubsequenceTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    auto array = test["array"].get<std::vector<int>>();
    auto expected = test["expected"].get<std::vector<int>>();

    auto result = longestIncreasingSubsequence(array);

    EXPECT_EQ(result, expected) << "Failed test case: " << test.dump(2);
  }
}
